ЛР1, вариант 2, ВТ-119 (Глушак, Нефедов, Сергеева)

egorglushak opened the pull request.

GOOD-Stuff commented:
1. не совсем понимаю ваш вопрос. Если вы по поводу прошлой работы, то там проблема с `typename` заключалась в том, что вы использовали её по отношению к уже известному компилятору типу (`list<int>`), но суть использования `typename` слегка другая . Проблема в том, что msvc не проверяет шаблоны полностью до тех пор, пока они не будут инстанцированы (вплоть до использования каждого метода). Возможно поэтому у вас подобные ошибки не отлавливаются.
> как мне не использовать typename

Вы можете явно объявить свой псевдоним на нужный вам тип (например ваш собственный *Iterator*): `typedef typename Map..::Iterator iter` (или типа того).
2. чтобы делать какие-либо утверждения по поводу скорости работы, необходимо сделать метрики (и желательно их предоставить). Пока реализацию не смотрел, ничего сказать не могу;
3. передавать пары, std::pair<KeyType, ValueType>, где KeyType и ValueType - параметры шаблона. Пример: `myMap <int, int> obj = { {0, 1}, {1, 1}, {2, 1} }`;
4. проще всего было бы вывести в консоль, но если на деле выборка большая (больше 100 значений), то возможно лучше записать их в файл. Основная цель задания - навыки оценки данных и определения, что из них должно быть ключом, а что значениями (способ идентификации массива данных). Т.е. вам нужно определить, что из полей должно быть ключом (уникальная последовательность, по которой можно максимально точно сопоставить значение), а что значением (не уникальные, чаще всего повторяемые данные). Нечто подобное у вас будет в будущем на курсах БД (базы данных).

GOOD-Stuff commented:
> Скорость работы словаря очень низкая

Пока на глаз могу сказать, что основная проблема в реализации. 
Словарь по своей сути - (красно-чёрное) дерево, вы же реализуете просто список. 
Список - медленная структура данных в принципе, особенно касательно реализации в С++.

GOOD-Stuff commented on lab_04/src/main.cpp:
```diff
@@ -0,0 +1,32 @@
+    struct main_struct { string name_str; string phone_str; };
+    int i = 0;
+
+    Map<int, main_struct> map;
```
Мне кажется, что делать ключом счётчик - плохая идея (для данного задания). 
Возможно я слегка переборщил с выборками, в принципе вы можете сократить их в половину.

GOOD-Stuff commented:
И ещё, если вы устанавливали себе *MinGW*, вы можете воспользоваться *Makefile*'ами, к-ые я прикладывал к предыдущим работам, чтобы проверить скомпилируется ли ваша работа у меня или нет (правда, возможно вам нужно будет подкрутить пути до компилятора *g++*).

GOOD-Stuff commented on lab_04/src/Map.h:
```diff
@@ -0,0 +1,291 @@
+		Node* ptr;//pointer to node
+	public:
+		Iterator() noexcept : ptr(nullptr) {};//const
+		Iterator(Node* root) noexcept : ptr(root) {};
```
**explicit**

GOOD-Stuff commented:
Хорошо, работу принимаю

egorglushak committed:
fix

egorglushak committed:
lab1

Malia753 committed:
Vector

zeRRo777 committed:
Queue

egorglushak committed:
2

egorglushak committed:
Merge branch 'pp' of https://github.com/egorglushak/ssw_lab_4sem into pp

GOOD-Stuff commented on src/Queue.cpp:
```diff
@@ -0,0 +1,130 @@
+#include <iostream>
+#include "Queue.h"
+#include "quick_sort.h"
```
Ошибка компиляции, такой файл отсутствует в вашем проекте

GOOD-Stuff commented on src/Queue.h:
```diff
@@ -0,0 +1,29 @@
+class Queue
+{
+private:
+	int* data;
+	int size;
+	int begin_id;
```
Инициализацию делайте в описании класса:
```C++
private:
  int* data {nullptr};
  int size {0};
  int begin_id {0};
```

GOOD-Stuff replied:
Конструктор в таком случае можно будет объявить как `default` и не реализовывать его:
`Queue() = default;`

GOOD-Stuff commented on src/Queue.cpp:
```diff
@@ -0,0 +1,130 @@
+}
+Queue::~Queue()
+{
+	if (data == nullptr) {
```
Ошибка в условии: если указатель `nullptr`, то его нельзя удалять (в принципе, удалять то и нечего, он ни на какую память в таком случае не указывает).

GOOD-Stuff commented on src/Queue.cpp:
```diff
@@ -0,0 +1,130 @@
+}
+void Queue::pop()
+{
+	if (size == 0)
+		cout << "queue is not founded";
```
В такой ситуации надо кидать исключение, в противном случае код отработает некорректно.

GOOD-Stuff commented on src/Queue.cpp:
```diff
@@ -0,0 +1,130 @@
+}
+int& Queue::peek()
+{
+	if (size == 0)
+		cout << "queue is not founded";
```
та же история, что и в `pop`

GOOD-Stuff commented on src/Queue.cpp:
```diff
@@ -0,0 +1,130 @@
+{
+	if (size == 0)
+		cout << "queue is not founded";
+	return data[size - 1];
```
Вершина в рамках очереди - её первый элемент.

GOOD-Stuff commented on src/Queue.cpp:
```diff
@@ -0,0 +1,130 @@
+}
+double Queue::get_average()
+{
+	if (size == 0)
+		cout << "vector is not founded";
```
А как вы представляете  себе исполнение программы? 
Ну вывели вы сообщение, а дальше что? У вас же программа продолжает исполняться, генерируя тем самым некорректное поведение, можно же было хотя бы банальный `return` вставить.
Но тут как и в случае с `peek` и `pop` вам придётся кидать исключение через `throw`:
`throw std::runtime_error("Vector is not founded");`.

GOOD-Stuff commented on src/Queue.cpp:
```diff
@@ -0,0 +1,130 @@
+{
+	Queue new_queue;
+
+	new_queue.size = size + queue.size;
+	new_queue.data = new int[new_queue.size];
```
Нарушаете инкапсуляцию. 
Эти параметры либо передаёте в конструктор, либо через соответствующие методы-сеттеры.

GOOD-Stuff commented on src/Stack.h:
```diff
@@ -0,0 +1,28 @@
+class Stack
+{
+private:
+	int* data;
+	int size;
```
та же история, что и у очереди

GOOD-Stuff commented on src/main.cpp:
```diff
@@ -0,0 +1,7 @@
+#include "Tests.h"
```
А этот файл где?

GOOD-Stuff commented on src/Vector.cpp:
```diff
@@ -0,0 +1,128 @@
+}
+Vector::~Vector()
+{
+	if (data == nullptr) {
```
Та же ошибка, что и у очереди.

GOOD-Stuff commented on src/Vector.cpp:
```diff
@@ -0,0 +1,128 @@
+{
+	if (size == 0)
+		cout << "vector is not founded";
+	if (--size % constant == 0) {
```
некорректно построили условие, из-за предыдущего вы лишь напортачите при работе с пустым вектором. 
кидайте исключения

GOOD-Stuff commented on src/Vector.cpp:
```diff
@@ -0,0 +1,128 @@
+}
+int& Vector::at(int index)
+{
+	if (size == 0)
+		cout << "vector has not elemet";
+	if (index < 0 || index > size - 1)
+		cout << "error";
```
ошибка, отсутствует `throw`.

GOOD-Stuff reviewed (commented):
### Замечания:
- ошибка компиляции: отсутствуют хедеры quick_sort и Test;
- некорректное условие в деструкторах: освобождаете пустые указатели;
- некорректное поведение контейнеров, когда они пусты;
- нарушение инкапсуляции в инд. задании.

За работу могу поставить 5.

egorglushak committed:
Исправления

Добавил недостающие файлы

egorglushak commented:
Добавил недостающие файлы.

egorglushak committed:
Lab2-буфер

zeRRo777 committed:
2 laba list

zeRRo777 committed:
Merge branch 'pp' of https://github.com/egorglushak/ssw_lab_4sem into pp

Malia753 committed:
laba 2 SList

egorglushak committed:
remove lab2

egorglushak committed:
lab2

Мы вчера скинули вторую лабораторную не в ту ветку, и из-за этого не смогли сделать пул реквест, сейчас его сделать уже нельзя, но лабораторную мы сдали в срок. Прикрепляю фалы второй лаб. раб. сюда, что бы вы заметили.

GOOD-Stuff commented on src/Tests.cpp:
```diff
@@ -0,0 +1,214 @@
+	cout << "Queue 1:\n\n" << my_queue_1 << "\n";
+	cout << "Queue 2:\n\n" << my_queue_2 << "\n";
+	cout << "Left - my queue, right - STL queue:\n\n";
+	cout << "Queue 1 == Queue 2: " << BoolToStr(my_queue_1 == my_queue_2) << " | " << BoolToStr(stl_queue_1 == stl_queue_2) << "\n";
```
Ну уж постыдились бы [копипастить](https://github.com/Overfirst/ssw_lab_4sem/blob/dev/src/ContainersTests.cpp#L107), свои то тесты уж можно накидать быстренько.

GOOD-Stuff reviewed (commented):
### Замечания:
- отсутствует исправление замечаний;
- копипаст;

За работу ставлю 5.
